cmd/ax/upgrade.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"archive/tar"
+	"bufio"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/google/go-github/github"
+	"github.com/kardianos/osext"
+)
+
+func getLatestReleaseData() *github.RepositoryRelease {
+	client := github.NewClient(nil)
+
+	release, _, err := client.Repositories.GetLatestRelease(context.Background(), "egnyte", "ax")
+	if err != nil {
+		fmt.Printf("Couldn't get a release information")
+		os.Exit(1)
+	}
+
+	return release
+}
+
+func getLatestAssetLink() string {
+	release := getLatestReleaseData()
+	var assetDownloadLink string
+
+	for _, element := range release.Assets {
+		assetLink := element.GetBrowserDownloadURL()
+		if strings.Contains(assetLink, runtime.GOOS) && strings.Contains(assetLink, runtime.GOARCH) {
+			assetDownloadLink = assetLink
+			break
+		}
+	}
+
+	return assetDownloadLink
+}
+
+func getLatestReleaseTag() string {
+	release := getLatestReleaseData()
+	return release.GetTagName()
+}
+
+func downloadFile(filepath string, url string) error {
+
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func upgradeVersion() error {
+	if latestVersion := getLatestReleaseTag(); version != latestVersion {
+		fmt.Printf("New version detected. Current version: %s. Latest version: %s\n", version, latestVersion)
+		inputScanner := bufio.NewScanner(os.Stdin)
+		fmt.Print("Do you want to upgrade an Ax to its latest version? (yes/no): ")
+		inputScanner.Scan()
+		text := inputScanner.Text()
+
+		if strings.ToLower(text) == "yes" || strings.ToLower(text) == "y" {
+			latestAssetURL := getLatestAssetLink()
+			downloadPath := "/tmp/egnyte_ax/ax.tar.gz"
+			extractionPath := "/tmp/egnyte_ax/"
+
+			if _, err := os.Stat(extractionPath); os.IsNotExist(err) {
+				os.Mkdir(extractionPath, 0755)
+			}
+
+			fmt.Println("Ax upgrade in progress...")
+
+			err := downloadFile(downloadPath, latestAssetURL)
+			if err != nil {
+				return err
+			}
+
+			file, err := os.Open(downloadPath)
+			if err != nil {
+				return err
+			}
+			defer os.Remove(downloadPath)
+
+			// extract the tar.gz file
+			err = extractTar(extractionPath, file)
+			if err != nil {
+				return err
+			}
+
+			// switch old binary with a new one
+			currentBinaryPath, _ := osext.Executable()
+			err = os.Rename(extractionPath+"ax", currentBinaryPath)
```

Comment:
Right, I suppose if you use TempFile and TempDir this may not work (rename won't work between different partitions). But you could copy the file and delete it afterwards.